enhancements/etcd/assisted-install.md

```diff
@@ -0,0 +1,221 @@
+the bootstrap node is repurposed into the third master node.
+
+## Challenges
+The installer currently does not provide an `install_type` key for the
+install-config. So no direct signal is passed from the installer to operator of
+the exact intent to perform a non standard install process.
+
+The result is the operator must make educated assumptions as to the install intent based
+on observations from the cluster itself.
+
+## Motivation
+Currently the `cluster-etcd-operator` does not tolerate less than 3 master
+nodes during install time.
+
+Reasons outlined for cluster with less than 3 master nodes as unsafe non HA.
+1. Violates operational best practices for etcd - unstable
+1. Allows non-HA configurations which we cannot support in production - unsafe and non-HA
+1. Allows a situation where we can get stuck unable to re-achieve quorum, resulting in cluster-death - unsafe, non-HA, non-production, unstable
+1. The combination of all these things makes the situation unsupportable. 
+
+To allow toleration of this check we provided an unsupportedConfigOverrides
+`useUnsupportedUnsafeNonHANonProductionUnstableEtcd`. As assisted installer is
+now moving to GA a supported toleration is required.
+
+## Goals
+
+1.  Provide a supported solution for assisted installer that will allow
+    `cluster-etcd-operator` to tolerate 2 master replicas as install complete.
+
+## Non-Goals
+
+1.  Support the current unsupportedConfigOverrides
+1.  Add support for less than 3 master node clusters
+
+## Currently Supported Functionality 
+Currently `cluster-etcd-operator` and OpenShift only supports clusters with
+three master nodes as install complete.
+
+### EnvVarController
+Currently the `EnvVarController` ensures that all of the data required for the
+etcd static pods exists. A condition exists within the `getEtcdEnvVars`[1]  function which checks if we have the expected master node count or if
+`useUnsupportedUnsafeNonHANonProductionUnstableEtcd` is populated otherwise
+return error which results in the controller being degraded and no static-pod
+spec installed on the master node for etcd.
+
+[1] https://github.com/openshift/cluster-etcd-operator/blob/be2b8d97ca048baecb051d288971a9f6db411457/pkg/etcdenvvar/etcd_env.go#L71
+
+## Proposed changes
+The following changes are proposed.
+
+In order for the operator to tolerate an unsafe install configuration all
+efforts must be taken to minimize the risk of tolerating user defined unsafe
+cluster. For example to save on costs user chooses 2 master replicas.
+
+1.  Observe master replica count from install-config.
+1.  Observe install status and only tolerate less than 3 masters until install
+    complete. After install status is complete the operator will go into a
+degraded state until the cluster has 3 or more master nodes.
+1.  Observe infrastructure and only tolerate `BareMetal` type. Assumption being
+    that assisted installer will GA with support for only `BareMetal` (true?)
```

Comment:
Note that IPI baremetal also uses the `platform: baremetal` type, so we can't assume that this toleration is always appropriate when that platform is selected.

AFAIK it is true that the assisted-installer GA will only support the baremetal platform, since that enables some components (hosted LB/DNS in particular) which are needed.

Reply:
Not sure it's 100% correct. with moving metal / networking part be part of SLO, I guess this platform will not have a lot of meaning. Also, we might want to provide more felxability to the user farther the road - provision metal3 but provide your network for example. Not sure what platform type will it be then.

Reply:
But for now yes, we use baremetal platform in the installer for a sake of manifests generation.

Reply:
It's not related to the SLO or day2, we can't declare bootstrap-complete until all 3 masters are up in the IPI case, because that will cause the bootstrap VM to be destroyed, when we might still be in the process of deploying 1 of the three masters - thus we can't tolerate a missing master in the IPI case.